routing/cross_mwm_index_graph.cpp

```diff
@@ -41,45 +80,106 @@ bool CrossMwmIndexGraph::IsTransition(Segment const & s, bool isOutgoing)
-    {
-      FillTransitionSegments(mappingPtr->m_segMapping, node.m_nodeId, s.GetMwmId(),
-                             transitionSegments.m_outgoing);
-    });
-    mappingPtr->m_crossContext.ForEachIngoingNode([&](IngoingCrossNode const & node)
-    {
-      FillTransitionSegments(mappingPtr->m_segMapping, node.m_nodeId, s.GetMwmId(),
-                             transitionSegments.m_ingoing);
-    });
-    auto const p = m_transitionCache.emplace(s.GetMwmId(), transitionSegments);
-    it = p.first;
-    CHECK(p.second, ("Mwm num id:", s.GetMwmId(), "has been inserted before. countryFile:",
-                     countryFile));
+    FillWholeMwmTransitionSegments(s.GetMwmId());
+    it = m_transitionCache.find(s.GetMwmId());
   }
+  CHECK(it != m_transitionCache.cend(),
+        ("Mwm ", s.GetMwmId(), "has not been downloaded. s:", s, ". isOutgoing", isOutgoing));
 
   if (isOutgoing)
     return it->second.m_outgoing.count(s) != 0;
   return it->second.m_ingoing.count(s) != 0;
 }
 
-void CrossMwmIndexGraph::GetTwins(Segment const & /* s */, std::vector<Segment> & /* twins */) const
+void CrossMwmIndexGraph::GetTwins(Segment const & s, bool isOutgoing, std::vector<Segment> & twins)
 {
+  CHECK(IsTransition(s, isOutgoing), ("IsTransition(", s, ",", isOutgoing, ") returns false."));
   // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
   // and if so to use it. If not, osrm cross mwm sections should be used.
-  NOTIMPLEMENTED();
+  auto const segMwmMapping = GetRoutingMapping(s.GetMwmId());
+  CHECK(segMwmMapping, ("Mwm", s.GetMwmId(), "has not been downloaded. s:", s, ". isOutgoing", isOutgoing));
+
+  twins.clear();
+  vector<string> const & neighboringMwm = segMwmMapping->m_crossContext.GetNeighboringMwmList();
+
+  for (string const & name : neighboringMwm)
+    FillWholeMwmTransitionSegments(m_numMwmIds->GetId(CountryFile(name)));
+
+  auto it = m_transitionCache.find(s.GetMwmId());
+  CHECK(it != m_transitionCache.cend(), ());
+
+  ms::LatLon const & latLog = isOutgoing ? GetLatLon(it->second.m_outgoing, s)
+                                         : GetLatLon(it->second.m_ingoing, s);
+  for (string const & name : neighboringMwm)
+  {
+    NumMwmId const numMwmId = m_numMwmIds->GetId(CountryFile(name));
+    TRoutingMappingPtr mappingPtr = GetRoutingMapping(numMwmId);
+    if (mappingPtr == nullptr)
+      continue; // mwm was not loaded.
+
+    if (isOutgoing)
+    {
+      mappingPtr->m_crossContext.ForEachIngoingNodeNearPoint(latLog, [&](IngoingCrossNode const & node){
+        AddTransitionSegment(mappingPtr->m_segMapping, node.m_nodeId, numMwmId, twins);
+      });
+    }
+    else
+    {
+      mappingPtr->m_crossContext.ForEachOutgoingNodeNearPoint(latLog, [&](OutgoingCrossNode const & node){
+        AddTransitionSegment(mappingPtr->m_segMapping, node.m_nodeId, numMwmId, twins);
+      });
+    }
+  }
+
+  my::SortUnique(twins);
```

Comment:
I don't know for sure. I haven't seen such cases. But taking into account that we may have duplicates in FillTransitionSegments(...) I guess it's possible that we can have them here.